### PR TITLE
refactor: reading environment variables; setting client cert source, api endpoints

### LINKED ISSUE #1871: Refactor the get_mtls_endpoint_and_cert_source function and consider updating the function name
For the function [get_mtls_endpoint_and_cert_source](https://github.com/googleapis/gapic-generator-python/blob/4b7701969f8e7c3ec5f23339cf7e9ad76f0dc757/gapic/templates/%25namespace/%25name_%25version/%25sub/services/%25service/client.py.j2#L220C9-L220C42), consider updating the name and refactor the code for better code readability and maintainability:

Suggested code change:
```python3
@classmethod
def validate_environment_variables(cls, use_client_cert, use_mtls_endpoint):
    if use_client_cert not in ("true", "false"):
        raise ValueError("Environment variable `GOOGLE_API_USE_CLIENT_CERTIFICATE` must be either `true` or `false`")
    if use_mtls_endpoint not in ("auto", "never", "always"):
        raise auth_exceptions.MutualTLSChannelError("Environment variable `GOOGLE_API_USE_MTLS_ENDPOINT` must be `never`, `auto` or `always`")

@classmethod
def get_client_cert_source(cls, client_options, use_client_cert):
    client_cert_source = None
    if use_client_cert == "true":
        if client_options.client_cert_source:
            client_cert_source = client_options.client_cert_source
        elif mtls.has_default_client_cert_source():
            client_cert_source = mtls.default_client_cert_source()
    return client_cert_source

@classmethod
def get_api_endpoint(cls, api_override, use_mtls_endpoint, client_cert_source):
    if api_override is not None:
        api_endpoint = api_override
    elif use_mtls_endpoint == "always" or (use_mtls_endpoint == "auto" and client_cert_source):
        api_endpoint = cls.DEFAULT_MTLS_ENDPOINT
    else:
       # return the default api endpoint
    return api_endpoint

@classmethod
def get_mtls_endpoint_and_cert_source(cls, client_options: Optional[GoogleClientOptions] = None) -> Tuple[str, Callable[[], Tuple[bytes, bytes]], Optional[str]]:
    if client_options is None:
        client_options = GoogleClientOptions()

    use_client_cert = os.getenv("GOOGLE_API_USE_CLIENT_CERTIFICATE", "false")
    use_mtls_endpoint = os.getenv("GOOGLE_API_USE_MTLS_ENDPOINT", "auto")

    cls.validate_environment_variables(use_client_cert, use_mtls_endpoint)

    client_cert_source = cls.get_client_cert_source(client_options, use_client_cert)
    api_endpoint = cls.get_api_endpoint(client_options.api_override, use_mtls_endpoint, client_cert_source)

    return api_endpoint, client_cert_source
```

The function name should also be updated to `get_api_endpoint_and_cert_source` since that is essentially what it's returning.

Also consider refactoring the code in here for consistency: [Ads template](https://github.com/googleapis/gapic-generator-python/blob/4b7701969f8e7c3ec5f23339cf7e9ad76f0dc757/gapic/ads-templates/%25namespace/%25name/%25version/%25sub/services/%25service/client.py.j2#L293)

### PR DESCRIPTION
**Pull request by ohmayr**
Fixes googleapis/gapic-generator-python#1871 🦕

### CODE REVIEW

**Review by parthea — APPROVED** [review]
LGTM, just minor observations

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/async_client.py.j2 [hunk 1]
```diff
@@ -191,6 +191,7 @@ class {{ service.async_client_name }}:
             client_info=client_info,
 
         )
+        self._api_endpoint = self._client._api_endpoint
```
[the remark this answers is not in the document]
Comment (ohmayr) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/async_client.py.j2:194]: Filed #1883

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 2]
```diff
@@ -323,15 +323,95 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
...
+            if self._client_options.api_endpoint is not None:
+                api_endpoint = self._client_options.api_endpoint
+            elif self._use_mtls_endpoint == "always" or (self._use_mtls_endpoint == "auto" and self._client_cert_source):
+                api_endpoint = self.DEFAULT_MTLS_ENDPOINT
+            else:
+                api_endpoint = self.DEFAULT_ENDPOINT
+
+            return api_endpoint
 
-        api_endpoint, client_cert_source_func = self.get_mtls_endpoint_and_cert_source(client_options)
```
[the remark this answers is not in the document]
Comment (ohmayr) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:332]: The alternative is not available as of now. I've filed a bug to deprecate this: #1885

**Review by parthea — APPROVED** [review]

**Review by vchudnov-g — CHANGES_REQUESTED** [review]
Thanks for working on this! Two general notes:

- You are not refactoring the `mtls_endpoint` function. You are not even touching it. So I would change the PR title to be more accurate, something like "inline logic for reading environment variables  and settingclient cert source, api endpoints"

- What is the value of inlining the logic that is in the mtls_endpoint function into the client `__init__` function? I would suggest making those inner functions take parameters and be module-level functions, and then you can test them directly to make sure the logic of each is as you desire.  Then in the test for the `__init__`  function, all you have test is that arbitrary values returned by (mocked versions of) those functions get properly used in the client. Doing this makes your code more modular and thus easier to test and reason about.

Happy to discuss more.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 3]
```diff
@@ -323,15 +323,95 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
...
-            client_options = client_options_lib.ClientOptions()
-        client_options = cast(client_options_lib.ClientOptions, client_options)
+        self._client_options = client_options
+        if isinstance(self._client_options, dict):
+            self._client_options = client_options_lib.from_dict(self._client_options)
+        if self._client_options is None:
+            self._client_options = client_options_lib.ClientOptions()
+        self._client_options = cast(client_options_lib.ClientOptions, self._client_options)
+
+        # validate the environment variables
```
[the remark this answers is not in the document]
Comment (ohmayr) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2]: addressed!

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 4]
```diff
@@ -323,15 +323,95 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
...
+        def validate_environment_variables():
```
[the remark this answers is not in the document]
Comment (ohmayr) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2]: addressed!

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 5]
```diff
@@ -323,15 +323,95 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
...
+        # Figure out the client cert source to use.
+        def get_client_cert_source():
+            """Return the client cert source used by the client.
+
+            The client cert source is determined in the following order:
+            (1) if `GOOGLE_API_USE_CLIENT_CERTIFICATE` environment variable is not "true", the
+            client cert source is None.
+            (2) if `client_options.client_cert_source` is provided, use the provided one; if the
+            default client cert source exists, use the default one; otherwise the client cert
+            source is None.
```
[the remark this answers is not in the document]
Comment (ohmayr) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2]: addressed!

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 6]
```diff
@@ -323,15 +323,95 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
...
+        self._client_cert_source = get_client_cert_source()
+
+
+        # Figure out which api endpoint to use.
+        def get_api_endpoint():
+            """Return the API endpoint used by the client.
+
+            The API endpoint is determined in the following order:
+            (1) if `client_options.api_endpoint` if provided, use the provided one.
+            (2) if `GOOGLE_API_USE_CLIENT_CERTIFICATE` environment variable is "always", use the
```
[the remark this answers is not in the document]
Comment (ohmayr) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2]: addressed!

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2 [hunk 7]
```diff
@@ -474,6 +478,77 @@ def test_{{ service.client_name|snake_case }}_get_mtls_endpoint_and_cert_source(
...
+        client = client_class(credentials=ga_credentials.AnonymousCredentials())
+        assert client._api_endpoint == client_class.DEFAULT_ENDPOINT
+
+    # If ClientOptions.api_endpoint is not set and GOOGLE_API_USE_MTLS_ENDPOINT="always",
+    # use the DEFAULT_MTLS_ENDPOINT as the api endpoint.
+    with mock.patch.dict(os.environ, {"GOOGLE_API_USE_MTLS_ENDPOINT": "always"}):
+        client = client_class(credentials=ga_credentials.AnonymousCredentials())
+        assert client._api_endpoint == client_class.DEFAULT_MTLS_ENDPOINT
+
+    # If ClientOptions.api_endpoint is not set and GOOGLE_API_USE_MTLS_ENDPOINT="auto" and default cert source doesn't exist,
```
[the remark this answers is not in the document]
Comment (ohmayr) [reply gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2]: addressed!

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2 [hunk 8]
```diff
@@ -474,6 +478,77 @@ def test_{{ service.client_name|snake_case }}_get_mtls_endpoint_and_cert_source(
...
+    # use the DEFAULT_ENDPOINT as the api endpoint.
+    with mock.patch.dict(os.environ, {"GOOGLE_API_USE_CLIENT_CERTIFICATE": "true"}):
+        with mock.patch('google.auth.transport.mtls.has_default_client_cert_source', return_value=False):
+            client = client_class(credentials=ga_credentials.AnonymousCredentials())
+            assert client._api_endpoint == client_class.DEFAULT_ENDPOINT
+
+    # If ClientOptions.api_endpoint is not set and GOOGLE_API_USE_MTLS_ENDPOINT="auto" and default cert source exists,
```
[the remark this answers is not in the document]
Comment (ohmayr) [reply gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2]: addressed!

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 9]
```diff
@@ -323,15 +323,95 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
...
+                api_endpoint = self._client_options.api_endpoint
+            elif self._use_mtls_endpoint == "always" or (self._use_mtls_endpoint == "auto" and self._client_cert_source):
+                api_endpoint = self.DEFAULT_MTLS_ENDPOINT
+            else:
+                api_endpoint = self.DEFAULT_ENDPOINT
+
+            return api_endpoint
 
-        api_endpoint, client_cert_source_func = self.get_mtls_endpoint_and_cert_source(client_options)
+        self._api_endpoint = get_api_endpoint()
```
[the remark this answers is not in the document]
Comment (ohmayr) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:393]: Yes, this PR is a part of alternative 2 in the doc. However, we don't want to deprecate the function until there's an alternative available i.e. users have another way of reading the api endpoint.

So we've decided to keep the function for now, and deprecate it once the TPC changes are in. That'll include access to the api endpoint via a property method.

I've filed a bug to deprecate this method: #1885

**Comment by ohmayr** [comment]
> * You are not refactoring the `mtls_endpoint` function. You are not even touching it. So I would change the PR title to be more accurate, something like "inline logic for reading environment variables  and settingclient cert source, api endpoints"

You're right! As we were working through it, we figured that it made more sense to not touch the function for now but have inline logic inside the client initialization so that the TPC changes in the design make more sense. Apart from that, the `get_mtls_endpoint_and_cert_source` function does more than what the name suggests, i.e. it isn't restricted to just getting the mtls endpoint. I've updated the PR title to reflect the changes.

> * What is the value of inlining the logic that is in the mtls_endpoint function into the client `__init__` function? I would suggest making those inner functions take parameters and be module-level functions, and then you can test them directly to make sure the logic of each is as you desire.  Then in the test for the `__init__`  function, all you have test is that arbitrary values returned by (mocked versions of) those functions get properly used in the client. Doing this makes your code more modular and thus easier to test and reason about.

Having these functions inside the `__init__` function provides us with encapsulation and better readability. By default, python doesn't have private methods i.e. any module level functions are accessible to the user (even the ones that we declare with an underscore to be private and not meant to be used). We probably don't want to support these functions right now.

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 10]
```diff
@@ -323,15 +323,95 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
...
+                Tuple[str, str]: returns the GOOGLE_API_USE_CLIENT_CERTIFICATE
+                    and the GOOGLE_API_USE_MTLS_ENDPOINT environment variables.
+
+            Raises:
+                ValueError: If GOOGLE_API_USE_CLIENT_CERTIFICATE is not
+                    any of ["true", "false"].
+                google.auth.exceptions.MutualTLSChannelError: If GOOGLE_API_USE_MTLS_ENDPOINT
+                    is not any of ["auto", "never", "always"].
+            """
+            use_client_cert = os.getenv("GOOGLE_API_USE_CLIENT_CERTIFICATE", "false")
```
[the remark this answers is not in the document]
Comment (ohmayr) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2]: addressed!

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 (hunk 9 above)
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:393]: Hmm, I see. But in the expected/normal workflow, users shouldn't need to call that function, right? I would think it still makes some sense to deprecate it now. To clarify, deprecation does not mean deletion: we keep the function but say that it is deprecated and will be removed at some point, as the client initialization no longer needs it.

More importantly, this argues for the benefits of not making the code inaccessible via inner functions, but moving it to independent function(s) that the user can access: that way, people who are calling the current function directly now can switch to call the new function directly.

**Review by vchudnov-g — APPROVED** [review]
Looks great! Thanks for doing this @ohmayr !

I have one suggestion to return a real boolean rather than a stringified one, and suggestions for some more checks in your existing tests. Everything else is cosmetic or nitty or deferrable.

**Comment by vchudnov-g** [comment]
Also, given how this PR wound up, maybe changing the title one more time: ` refactor logic for reading environment variables, and for setting client cert source and api endpoints`

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2 [hunk 11]
```diff
@@ -100,6 +100,56 @@ def test__get_default_mtls_endpoint():
...
+    with mock.patch.dict(os.environ, {"GOOGLE_API_USE_CLIENT_CERTIFICATE": "true"}):
+        assert {{ service.client_name }}._read_environment_variables() == ("true", "auto")
+    
+    with mock.patch.dict(os.environ, {"GOOGLE_API_USE_CLIENT_CERTIFICATE": "Unsupported"}):
+        with pytest.raises(ValueError) as excinfo:
+            {{ service.client_name }}._read_environment_variables()
+
+    assert str(excinfo.value) == "Environment variable `GOOGLE_API_USE_CLIENT_CERTIFICATE` must be either `true` or `false`"
+
+    with mock.patch.dict(os.environ, {"GOOGLE_API_USE_MTLS_ENDPOINT": "never"}):
```
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2:115]: Could you add a stanza for this case: `with mock.patch.dict(os.environ, {"GOOGLE_API_USE_MTLS_ENDPOINT": "never"}):`? (You're covering the default and the unrecognized value, but what about the explictly set `auto`?

And similarly for `with mock.patch.dict(os.environ, {"GOOGLE_API_USE_CLIENT_CERTIFICATE": false"}):` explicitly

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2 [hunk 12]
```diff
@@ -474,6 +528,79 @@ def test_{{ service.client_name|snake_case }}_get_mtls_endpoint_and_cert_source(
...
+    {% endif %}
+])
+@mock.patch.object({{ service.client_name }}, "DEFAULT_ENDPOINT", modify_default_endpoint({{ service.client_name }}))
+{% if 'grpc' in opts.transport %}
+@mock.patch.object({{ service.async_client_name }}, "DEFAULT_ENDPOINT", modify_default_endpoint({{ service.async_client_name }}))
+{% endif %}
+def test_{{ service.client_name|snake_case }}_client_api_endpoint(client_class):
+    mock_client_cert_source = client_cert_source_callback
+
+    # If ClientOptions.api_endpoint is set and GOOGLE_API_USE_CLIENT_CERTIFICATE="true",
```
[the remark this answers is not in the document]
Comment (vchudnov-g) [reply gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2:560]: ping

**Review by vchudnov-g — APPROVED** [review]
Looks good. I pinged you on a couple of comments from the last review that would still be a good idea.

And also, please see my comments about updating the PR title to be more accurate given our changes.

Thank you for doing this!

**Review by vchudnov-g — APPROVED** [review]

**Pull request merged**